packages/plugin-drizzle/src/index.ts

```diff
@@ -1,5 +1,223 @@
+
+  if (!context[DRIZZLE_PROPERTY]) {
+    throw new DrizzleStorageError(
+      "drizzle storage is not available. Did you register the plugin?",
+    );
+  }
+
+  return context[DRIZZLE_PROPERTY];
+}
+
+export interface DrizzleStorageOptions<
+  TQueryResult extends PgQueryResultHKT,
+  TFullSchema extends Record<string, unknown> = Record<string, never>,
+  TSchema extends
+    TablesRelationalConfig = ExtractTablesWithRelations<TFullSchema>,
+> {
+  db: PgDatabase<TQueryResult, TFullSchema, TSchema>;
+  persistState?: boolean;
+  indexerName?: string;
+  schema?: Record<string, unknown>;
+  idColumn?: string;
+}
+
+export function drizzleStorage<
+  TFilter,
+  TBlock,
+  TQueryResult extends PgQueryResultHKT,
+  TFullSchema extends Record<string, unknown> = Record<string, never>,
+  TSchema extends
+    TablesRelationalConfig = ExtractTablesWithRelations<TFullSchema>,
+>({
+  db,
+  persistState: enablePersistence = true,
+  indexerName,
+  schema = {},
+  idColumn = "id",
+}: DrizzleStorageOptions<TQueryResult, TFullSchema, TSchema>) {
+  return defineIndexerPlugin<TFilter, TBlock>((indexer) => {
+    const tableNames = Object.keys(schema ?? db._.schema);
+
+    indexer.hooks.hook("run:before", async () => {
+      await withTransaction(db, async (tx) => {
+        await initializeReorgRollbackTable(tx);
+        if (enablePersistence) {
+          await initializePersistentState(tx);
+        }
+      });
+    });
+
+    indexer.hooks.hook("connect:before", async ({ request }) => {
+      if (!enablePersistence) {
+        return;
+      }
+
+      await withTransaction(db, async (tx) => {
+        const { cursor, filter } = await getState<
+          TFilter,
+          TQueryResult,
+          TFullSchema,
+          TSchema
+        >({
+          tx,
+          indexerName,
+        });
+        if (cursor) {
+          request.startingCursor = cursor;
+        }
+        if (filter) {
+          request.filter[1] = filter;
+        }
+      });
+    });
+
+    indexer.hooks.hook("connect:after", async ({ request }) => {
+      // On restart, we need to invalidate data for blocks that were processed but not persisted.
+      const cursor = request.startingCursor;
+
+      if (!cursor) {
+        return;
+      }
+
+      await withTransaction(db, async (tx) => {
+        // TODO: Implement invalidate for drizzle/reorg table maybe?
+        // await invalidate(db, cursor, tables);
+
+        if (enablePersistence) {
+          await invalidateState({ tx, cursor, indexerName });
+        }
+      });
+    });
+
+    indexer.hooks.hook("connect:factory", async ({ request, endCursor }) => {
+      if (!enablePersistence) {
+        return;
+      }
+      await withTransaction(db, async (tx) => {
+        if (endCursor && request.filter[1]) {
+          await persistState({
+            tx,
+            endCursor,
+            filter: request.filter[1],
+            indexerName,
+          });
+        }
+      });
```

Comment:
I think `connect:factory` should be executed in the same transaction as `handler:middlleware`.  Access it through the context.